File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/ai/harness/maven/rules/EmptyDirectoryPlaceholdersRule.java
package ai.harness.maven.rules;

import ai.harness.maven.Finding;

import tools.jackson.databind.JsonNode;
import org.apache.maven.plugin.MojoExecutionException;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Collection;
import java.util.List;
import java.util.stream.Stream;

/**
 * Rule that requires .gitkeep or real files in empty directories.
 */
public enum EmptyDirectoryPlaceholdersRule implements HarnessCheckRule {
    INSTANCE;

    @Override
    public String category() {
        return "emptyDirectoryPlaceholders";
    }
    private static final String CATEGORY = "emptyDirectoryPlaceholders";

    @Override
    public boolean applies(JsonNode manifest) {
        return HarnessCheckHelper.applies(manifest, CATEGORY);
    }

    @Override
    public Collection<Finding> validate(Path root, JsonNode manifest) throws MojoExecutionException {
        final JsonNode catNode = manifest.get(CATEGORY);
        final String severity = HarnessCheckHelper.getSeverity(manifest, CATEGORY);
        return HarnessCheckHelper.extractPaths(catNode.get("parameters").get("directories")).stream()
                .flatMap(dir -> validateKeepFile(root, dir, severity).stream())
                .toList();
    }

    private List<Finding> validateKeepFile(Path root, String dir, String severity) {
        final Path dirPath = root.resolve(dir);
        return Stream.of(dirPath)
                .filter(p -> HarnessCheckHelper.isSafeDirectory(root, p))
                .flatMap(p -> {
                    try (final Stream<Path> stream = Files.list(p)) {
                        final List<Path> realFiles = stream
                                .filter(f -> !Files.isSymbolicLink(f))
                                .filter(f -> !f.getFileName().toString().equals(".gitkeep"))
                                .toList();
                        return realFiles.isEmpty() && !HarnessCheckHelper.isSafeRegularFile(root, p.resolve(".gitkeep"))
                                ? Stream.of(new Finding(severity, CATEGORY, "empty directory must keep placeholder or real files: " + dir))
                                : Stream.empty();
                    } catch (IOException e) {
                        return Stream.empty();
                    }
                })
                .toList();
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/ai/harness/gradle/rules/EmptyDirectoryPlaceholdersRule.kt
package ai.harness.gradle.rules

import ai.harness.gradle.HarnessPsiResults.Finding
import ai.harness.gradle.HarnessCheck
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.nio.file.Path
import kotlin.io.path.div
import kotlin.io.path.exists
import kotlin.io.path.isDirectory
import kotlin.io.path.listDirectoryEntries
import kotlin.io.path.name

/**
 * Rule that requires .gitkeep or real files in empty directories.
 */
object EmptyDirectoryPlaceholdersRule : HarnessCheckRule() {
    override val category: String = "emptyDirectoryPlaceholders"
    override fun applies(manifest: JsonObject): Boolean =
        manifest[category]?.jsonObject?.get("enabled")?.jsonPrimitive?.contentOrNull?.toBoolean() ?: true

    override fun validate(
        manifest: JsonObject,
        root: Path,
    ): Collection<Finding> {
        val severity = HarnessCheck.severityOf(manifest, category)
        val catObj = manifest[category]?.jsonObject
        val parametersObj = catObj?.get("parameters")?.jsonObject
        return when {
            catObj == null || parametersObj == null -> {
                emptyList()
            }

            else -> {
                val directories = HarnessCheck.stringArrayFrom(parametersObj, "directories")
                directories
                    .filter { dirPath ->
                        val dir = root / dirPath
                        when {
                            !dir.isDirectory() -> {
                                false
                            }

                            else -> {
                                /**
                                 * Directory enumeration may fail; return empty on silent fallback to treat as containing no real files.
                                 */
                                val realFiles =
                                    try {
                                        dir.listDirectoryEntries().filter { entry -> entry.name != ".gitkeep" }
                                    } catch (e: Exception) {
                                        emptyList()
                                    }
                                realFiles.isEmpty() && !(root / "$dirPath/.gitkeep").exists()
                            }
                        }
                    }.map { dirPath ->
                        Finding(
                            severity,
                            category,
                            "empty directory must keep placeholder or real files: $dirPath",
                        )
                    }
            }
        }
    }

    override fun renderPsiFindings(
        findings: List<ai.harness.gradle.PsiFinding>,
        manifest: JsonObject,
    ): Collection<Finding> = emptyList()

    override fun findPsiFindings(
        file: Path,
        root: Path,
        psiFactory: org.jetbrains.kotlin.psi.KtPsiFactory?,
    ): List<ai.harness.gradle.PsiFinding> = emptyList()
}
